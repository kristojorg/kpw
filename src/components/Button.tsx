/** @jsx jsx */
import { jsx, Styled } from "theme-ui";
import { Button as BaseButton } from "reakit";
import * as React from "react";
import { ButtonVariants, VariantProp } from "../interfaces";
import Link from "./Link";
import { Box, PolymorphicComponentProps } from "./PolymorphicBox";

type Variant = VariantProp<ButtonVariants>;
type ButtonOwnProps = { variant?: Variant; disabled?: boolean };
type ButtonProps<E extends React.ElementType> = PolymorphicComponentProps<
  E,
  ButtonOwnProps
>;

const defaultComponent = BaseButton;

/**
 * renders anything with button styles from the theme. Pass an
 * "as" prop to control the backing component
 */
function Button<E extends React.ElementType = typeof defaultComponent>({
  variant = "primary",
  ...props
}: ButtonProps<E>): JSX.Element {
  return (
    <Box
      as={defaultComponent}
      sx={{ variant: `buttons.${variant}` }}
      {...props}
    />
  );
}

type NavButtonProps = React.ComponentProps<typeof Link> & ButtonOwnProps;
export function NavButton(props: NavButtonProps) {
  return <Button as={Link} {...props} />;
}

type AnchorButtonProps = React.ComponentProps<typeof Styled.a> & ButtonOwnProps;
export function AnchorButton(props: AnchorButtonProps) {
  return <Button as="a" {...props} />;
}
export default Button;
