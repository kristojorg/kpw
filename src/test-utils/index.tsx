import * as React from "react";
import { render } from "@testing-library/react";
import "@testing-library/jest-dom";
import { ThemeProvider } from "theme-ui";
import theme from "../theme";
import ContextProvider from "../components/context/ContextProvider";
import Adapter from "enzyme-adapter-react-16";
import { configure } from "enzyme";
import library from "./fixtures/library";
import { Router } from "react-router-dom";
import { createMemoryHistory } from "history";
import BasicAuthPlugin from "../auth/basicAuthPlugin";
import buildStore from "opds-web-client/lib/store";
import { State } from "opds-web-client/lib/state";
import { LibraryData } from "../interfaces";
import "./mockScrollTo";
import * as fixtures from "./fixtures";
import ActionsCreator from "opds-web-client/lib/actions";
import DataFetcher from "opds-web-client/lib/DataFetcher";
import { adapter } from "opds-web-client/lib/OPDSDataAdapter";
import serializer from "jest-emotion";

expect.addSnapshotSerializer(serializer);

/**
 * uncomment this if you would like console errors to to
 * error the test. Useful if you don't know where a console
 * error is coming from.
 */
// const consoleErrorSpy = jest.spyOn(global.console, "error");
// consoleErrorSpy.mockImplementation((msg, ...opts) => {
//   console.warn(msg, ...opts);
//   throw new Error(msg);
// });

export { fixtures };

configure({ adapter: new Adapter() });

/**
 * mock out the window.URL.createObjectURL since it isn't
 * available on jsdom
 */
const mockCreateObjectURL = jest.fn();
global["URL"].createObjectURL = mockCreateObjectURL;

/**
 * We create the actions and fetcher here so that they can
 * be imported by our test file and spied on using:
 * jest.spyOn(actions, "fetchSearchDescription")
 * and then asserted against
 */
export const fetcher = new DataFetcher({ adapter });
export const actions = new ActionsCreator(fetcher);

type CustomRenderOptions = Parameters<typeof render>[1] & {
  route?: string;
  initialState?: State;
  library?: LibraryData;
};
const customRender = (ui: any, options?: CustomRenderOptions) => {
  const pathFor = jest
    .fn()
    .mockImplementation((collectionUrl?: string, bookUrl?: string) =>
      `/${collectionUrl}` + bookUrl ? `/${bookUrl}` : ""
    );

  const store = buildStore(options?.initialState, [BasicAuthPlugin], pathFor);

  // spy on dispatch by default
  const origDispatch = store.dispatch;
  const mockDispatch: jest.Mock<typeof origDispatch> = jest
    .fn()
    .mockImplementation(origDispatch);
  store.dispatch = mockDispatch as typeof origDispatch;
  // const dispatch = jest.spyOn(store, "dispatch");

  const history = createMemoryHistory({
    initialEntries: [options?.route ?? "/"]
  });
  const AllTheProviders = ({ children }) => {
    return (
      <Router history={history}>
        <ThemeProvider theme={theme}>
          <ContextProvider
            library={options?.library ?? library}
            shortenUrls
            helmetContext={{}}
            initialState={options?.initialState}
            store={store}
            fetcher={fetcher}
            actions={actions}
          >
            {children}
          </ContextProvider>
        </ThemeProvider>
      </Router>
    );
  };

  return {
    ...render(ui, { wrapper: AllTheProviders, ...options }),
    // we pass our mocks along so they can be used in assertions
    store,
    history,
    dispatch: mockDispatch
  };
};

// re-export everything
export * from "@testing-library/react";

// override render method
export { customRender as render };
