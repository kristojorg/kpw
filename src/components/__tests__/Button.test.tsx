import * as React from "react";
import { render } from "../../test-utils";
import Button, { NavButton, AnchorButton } from "../Button";
import userEvent from "@testing-library/user-event";

describe("variants", () => {
  test("primary (default)", () => {
    const node = render(<Button>child</Button>);
    const button = node.getByText("child");
    expect(button).toMatchSnapshot();
  });

  test("flat", () => {
    const node = render(<Button variant="flat">child</Button>);
    const button = node.getByText("child");
    expect(button).toMatchSnapshot();
  });
  test("accent", () => {
    const node = render(<Button variant="accent">child</Button>);
    const button = node.getByText("child");
    expect(button).toMatchSnapshot();
  });
});
test("disabled style", () => {
  const node = render(<Button disabled>child</Button>);
  const button = node.getByText("child");
  expect(button).toBeDisabled();
  expect(button).toMatchSnapshot();
});
test("NavButton renders correct element", () => {
  const node = render(<NavButton to="/somewhere">child</NavButton>);
  const button = node.getByText("child");
  expect(button).toHaveAttribute("href", "/somewhere");
  expect(button).toMatchSnapshot();
  // make sure clicking it actually tries to route you home
  userEvent.click(button);
  expect(node.history.location.pathname).toBe("/somewhere");
});
test("LinkButton renders correct element", () => {
  const node = render(
    <AnchorButton href="http://some.com/external">child</AnchorButton>
  );
  const button = node.getByText("child");
  expect(button).toHaveAttribute("href", "http://some.com/external");
  expect(button).toMatchSnapshot();
});
