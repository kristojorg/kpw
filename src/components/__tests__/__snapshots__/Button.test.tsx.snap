// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LinkButton renders correct element 1`] = `
.emotion-0 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  text-align: center;
  line-height: inherit;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-size: inherit;
  font-weight: 700;
  margin: 0;
  padding-left: 16px;
  padding-right: 16px;
  padding-top: 4px;
  padding-bottom: 4px;
  border: 0;
  border-radius: 4px;
  cursor: pointer;
  text-transform: uppercase;
  -webkit-letter-spacing: 0.05em;
  -moz-letter-spacing: 0.05em;
  -ms-letter-spacing: 0.05em;
  letter-spacing: 0.05em;
  background-color: var(--theme-ui-colors-primary,#0467a6);
  color: white;
  fill: white;
}

.emotion-0:focus,
.emotion-0:hover {
  background-color: #03588d;
}

.emotion-0:active {
  background-color: #034874;
}

.emotion-0:disabled {
  background-color: var(--theme-ui-colors-disabled,#B3B9BE);
  cursor: default;
}

<a
  class="emotion-0"
  href="http://some.com/external"
>
  child
</a>
`;

exports[`NavButton renders correct element 1`] = `
.emotion-0 {
  -webkit-text-decoration: none;
  text-decoration: none;
  color: inherit;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  text-align: center;
  line-height: inherit;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-size: inherit;
  font-weight: 700;
  margin: 0;
  padding-left: 16px;
  padding-right: 16px;
  padding-top: 4px;
  padding-bottom: 4px;
  border: 0;
  border-radius: 4px;
  cursor: pointer;
  text-transform: uppercase;
  -webkit-letter-spacing: 0.05em;
  -moz-letter-spacing: 0.05em;
  -ms-letter-spacing: 0.05em;
  letter-spacing: 0.05em;
  background-color: var(--theme-ui-colors-primary,#0467a6);
  color: white;
  fill: white;
}

.emotion-0:focus,
.emotion-0:hover {
  background-color: #03588d;
}

.emotion-0:active {
  background-color: #034874;
}

.emotion-0:disabled {
  background-color: var(--theme-ui-colors-disabled,#B3B9BE);
  cursor: default;
}

<a
  class="emotion-0"
  href="/somewhere"
>
  child
</a>
`;

exports[`disabled style 1`] = `
.emotion-0 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  text-align: center;
  line-height: inherit;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-size: inherit;
  font-weight: 700;
  margin: 0;
  padding-left: 16px;
  padding-right: 16px;
  padding-top: 4px;
  padding-bottom: 4px;
  border: 0;
  border-radius: 4px;
  cursor: pointer;
  text-transform: uppercase;
  -webkit-letter-spacing: 0.05em;
  -moz-letter-spacing: 0.05em;
  -ms-letter-spacing: 0.05em;
  letter-spacing: 0.05em;
  background-color: var(--theme-ui-colors-primary,#0467a6);
  color: white;
  fill: white;
}

.emotion-0:focus,
.emotion-0:hover {
  background-color: #03588d;
}

.emotion-0:active {
  background-color: #034874;
}

.emotion-0:disabled {
  background-color: var(--theme-ui-colors-disabled,#B3B9BE);
  cursor: default;
}

<button
  aria-disabled="true"
  class="emotion-0"
  disabled=""
  style="pointer-events: none;"
  type="button"
>
  child
</button>
`;

exports[`variants accent 1`] = `
.emotion-0 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  text-align: center;
  line-height: inherit;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-size: inherit;
  font-weight: 700;
  margin: 0;
  padding-left: 16px;
  padding-right: 16px;
  padding-top: 4px;
  padding-bottom: 4px;
  border: 0;
  border-radius: 4px;
  cursor: pointer;
  text-transform: uppercase;
  -webkit-letter-spacing: 0.05em;
  -moz-letter-spacing: 0.05em;
  -ms-letter-spacing: 0.05em;
  letter-spacing: 0.05em;
  background-color: var(--theme-ui-colors-accent,#D00854);
  color: white;
  fill: white;
}

.emotion-0:focus,
.emotion-0:hover {
  background-color: #b7074a;
}

.emotion-0:active {
  background-color: #9f0640;
}

.emotion-0:disabled {
  background-color: var(--theme-ui-colors-disabled,#B3B9BE);
  cursor: default;
}

<button
  class="emotion-0"
  type="button"
>
  child
</button>
`;

exports[`variants flat 1`] = `
.emotion-0 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  text-align: center;
  line-height: inherit;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-size: inherit;
  font-weight: 700;
  margin: 0;
  padding-left: 16px;
  padding-right: 16px;
  padding-top: 4px;
  padding-bottom: 4px;
  border: 0;
  border-radius: 4px;
  cursor: pointer;
  text-transform: uppercase;
  -webkit-letter-spacing: 0.05em;
  -moz-letter-spacing: 0.05em;
  -ms-letter-spacing: 0.05em;
  letter-spacing: 0.05em;
  background-color: white;
  color: blues.dark;
}

.emotion-0:focus,
.emotion-0:hover {
  background-color: rgba(4,103,166,0.2);
}

.emotion-0:active {
  background-color: rgba(4,103,166,0.4);
}

.emotion-0:disabled {
  color: var(--theme-ui-colors-grey,#B3B9BE);
  cursor: default;
}

<button
  class="emotion-0"
  type="button"
>
  child
</button>
`;

exports[`variants primary (default) 1`] = `
.emotion-0 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  text-align: center;
  line-height: inherit;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-size: inherit;
  font-weight: 700;
  margin: 0;
  padding-left: 16px;
  padding-right: 16px;
  padding-top: 4px;
  padding-bottom: 4px;
  border: 0;
  border-radius: 4px;
  cursor: pointer;
  text-transform: uppercase;
  -webkit-letter-spacing: 0.05em;
  -moz-letter-spacing: 0.05em;
  -ms-letter-spacing: 0.05em;
  letter-spacing: 0.05em;
  background-color: var(--theme-ui-colors-primary,#0467a6);
  color: white;
  fill: white;
}

.emotion-0:focus,
.emotion-0:hover {
  background-color: #03588d;
}

.emotion-0:active {
  background-color: #034874;
}

.emotion-0:disabled {
  background-color: var(--theme-ui-colors-disabled,#B3B9BE);
  cursor: default;
}

<button
  class="emotion-0"
  type="button"
>
  child
</button>
`;
